src/Lykke.Service.PayInternal.Services/BtcTransferRequestService.cs

```diff
@@ -0,0 +1,428 @@
+        }
+        #endregion
+
+        #region public
+
+        /// <summary>
+        /// Tries to execute a crosswise transfer request from many sources to many destinations.
+        /// </summary>
+        /// <param name="transferRequest">The transfer request.</param>
+        /// <returns>The updated transfer request OR transfer request with error.</returns>
+        public async Task<ITransferRequest> CreateTransferCrosswiseAsync(ITransferRequest transferRequest)
+        {
+            // We need an independent copy of the request here, for some methods in current implementation change the inner state of the given object.
+            // For instance, the BtcTransferRequest class' constructor reuses the transaction list it gets.
+            var transferRequestUpdated = transferRequest.DeepCopy();
+
+            // First of all, we need to verify the source addresses to belong to the Merchant.
+            var sourcesCheckupResult = await CheckupSourcesAndWalletsAsync(transferRequestUpdated);
+            if (sourcesCheckupResult != null)
+                return sourcesCheckupResult;
+
+            // Here would be no additional validations. If transaction fails, we will know it from the underlying BitcoinApi service.
+
+            // Save the initial state of transfer
+            await _transferRepository.SaveAsync(transferRequestUpdated);
+            await _transferRequestPublisher.PublishAsync(transferRequestUpdated);
+
+            // The main work
+            foreach (var tran in transferRequestUpdated.TransactionRequests)
+            {
+                var sourceAddresses = (from s in tran.SourceAmounts
+                                       select new ToOneAddress(s.Address, s.Amount)).ToList();
+                // Please, note: feeRate = 0 and fixedFee = 0 here
+                var currentResult = await _bitcointApiClient.TransactionMultipleTransfer(null, tran.DestinationAddress, tran.Currency, 0, 0, sourceAddresses);
+
+                if (currentResult == null || currentResult.HasError)
+                {
+                    // Please, note: after the first faulty transaction we abort the further execution.
+                    // BUT money that we had already transfered will not go back to source address(es).
+                    // To make the whole transfer request atomic we need to implement such feature in
+                    // BitcoinApi service at first.
+                    return await PackFaultyResult(currentResult, tran, transferRequestUpdated);
+                }
+                else
+                {
+                    tran.TransactionHash = currentResult.Transaction.Hash;
+                }
+
+                await _transferRepository.SaveAsync(transferRequestUpdated);
+                await _transferRequestPublisher.PublishAsync(transferRequestUpdated);
+            }
+
+            transferRequestUpdated.TransferStatus = TransferStatus.InProgress;
```

Comment:
It seems here is some misunderstanding, it is unlikely usual behaviour to update the state of inbound object. There might be some reasons.